### PR TITLE
Multilayer selection drop support

### PR DESCRIPTION
**Pull request by roya0045**
## Description

This is a small improvement I had laying around, feel free to close it if I already have too many PR pending review.

This simply add drop support for layer in the layer selection widget to more efficiently select the desired elements. Especially useful in big project of with duplicated layer names.
drag drop_multi

### CODE REVIEW

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 1]
```diff
@@ -137,10 +142,16 @@ class GUI_EXPORT QgsProcessingMultipleSelectionPanelWidget : public QgsPanelWidg
 
     QList< QStandardItem * > currentItems();
 
-
     void populateList( const QVariantList &availableOptions, const QVariantList &selectedOptions );
 
     friend class TestProcessingGui;
+    bool mDragActive = false;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:148]: Tracking this manually should not be required -- it's all handled for us by Qt.

Comment (roya0045) [reply src/gui/processing/qgsprocessingmultipleselectiondialog.h]: Thanks for the clarification, I wasn't sure what was needed to properly handle everything. I removed the dragLeaveEvent as I guess it doesn't matter much.

**Comment by roya0045** [comment]
mingw build failure is unrelated.

**Comment by roya0045** [comment]
fine for now

**Comment by roya0045** [comment]
Works as intended.

**Comment by roya0045** [comment]
ready for review

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 2]
```diff
@@ -196,6 +198,65 @@ void QgsProcessingMultipleSelectionPanelWidget::populateList( const QVariantList
...
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
+  if ( !indexes.isEmpty() )
+  {
+    // dragged an acceptable layer, phew
+    event->setDropAction( Qt::CopyAction );
+    event->accept();
+    update();
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:235]:
```suggestion
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 3]
```diff
@@ -196,6 +198,65 @@ void QgsProcessingMultipleSelectionPanelWidget::populateList( const QVariantList
...
+  }
+}
+
+void QgsProcessingMultipleSelectionPanelWidget::dropEvent( QDropEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:244]:
```suggestion
  const QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 4]
```diff
@@ -196,6 +198,65 @@ void QgsProcessingMultipleSelectionPanelWidget::populateList( const QVariantList
...
+    event->setDropAction( Qt::CopyAction );
+    event->accept();
+
+    for ( const int i : indexes )
+    {
+      mModel->item( i )->setCheckState( Qt::Checked );
+    }
+    emit selectionChanged();
+  }
+  update();
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:258]:
```suggestion
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 5]
```diff
@@ -341,6 +402,66 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
+  if ( !indexes.isEmpty() )
+  {
+    // dragged an acceptable layer, phew
+    event->setDropAction( Qt::CopyAction );
+    event->accept();
+    update();
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:439]:
```suggestion
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 6]
```diff
@@ -341,6 +402,66 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+  }
+}
+
+void QgsProcessingMultipleInputPanelWidget::dropEvent( QDropEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:448]:
```suggestion
  const QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 7]
```diff
@@ -341,6 +402,66 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+    event->setDropAction( Qt::CopyAction );
+    event->accept();
+
+    for ( const int i : indexes )
+    {
+      mModel->item( i )->setCheckState( Qt::Checked );
+    }
+    emit selectionChanged();
+  }
+  update();
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:462]:
```suggestion
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 8]
```diff
@@ -246,8 +255,18 @@ class GUI_EXPORT QgsProcessingMultipleInputPanelWidget : public QgsProcessingMul
     void addFiles();
     void addDirectory();
 
+  protected:
+    void dragEnterEvent( QDragEnterEvent *event ) override SIP_SKIP;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:259]:
```suggestion
    void dragEnterEvent( QDragEnterEvent *event ) override;
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 9]
```diff
@@ -246,8 +255,18 @@ class GUI_EXPORT QgsProcessingMultipleInputPanelWidget : public QgsProcessingMul
...
+    void dropEvent( QDropEvent *event ) override SIP_SKIP;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:260]:
```suggestion
    void dropEvent( QDropEvent *event ) override;
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 10]
```diff
@@ -246,8 +255,18 @@ class GUI_EXPORT QgsProcessingMultipleInputPanelWidget : public QgsProcessingMul
...
+
   private:
+    bool mDragActive = false;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:263]:
```suggestion
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 11]
```diff
@@ -124,6 +124,10 @@ class GUI_EXPORT QgsProcessingMultipleSelectionPanelWidget : public QgsPanelWidg
     QStandardItemModel *mModel = nullptr;
     //! Value formatter
     std::function< QString( const QVariant & )> mValueFormatter;
+
+    void dragEnterEvent( QDragEnterEvent *event ) override SIP_SKIP;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:128]:
```suggestion
    void dragEnterEvent( QDragEnterEvent *event ) override;
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.h [hunk 12]
```diff
@@ -124,6 +124,10 @@ class GUI_EXPORT QgsProcessingMultipleSelectionPanelWidget : public QgsPanelWidg
...
+    void dropEvent( QDropEvent *event ) override SIP_SKIP;
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.h:129]:
```suggestion
    void dropEvent( QDropEvent *event ) override;
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@roya0045 can you update the sip files here please?

**Comment by roya0045** [comment]
> @roya0045 can you update the sip files here please?

Thanks for the review, should be done.

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 13]
```diff
@@ -196,6 +198,63 @@ void QgsProcessingMultipleSelectionPanelWidget::populateList( const QVariantList
...
+  return indexes;
+}
+
+void QgsProcessingMultipleSelectionPanelWidget::dragEnterEvent( QDragEnterEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
+  if ( !indexes.isEmpty() )
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:230]:
```suggestion
  const QList< int> indexes = existingMapLayerFromMimeData( event->mimeData() );
  if ( !indexes.isEmpty() )
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 14]
```diff
@@ -196,6 +198,63 @@ void QgsProcessingMultipleSelectionPanelWidget::populateList( const QVariantList
...
+    event->accept();
+  }
+}
+
+void QgsProcessingMultipleSelectionPanelWidget::dropEvent( QDropEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  const QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:243]:
```suggestion
  const QList< int> indexes = existingMapLayerFromMimeData( event->mimeData() );
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 15]
```diff
@@ -341,6 +400,64 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+  QList<int> indexes;
+  for ( const QgsMimeDataUtils::Uri &u : uriList )
+  {
+    // is this uri from the current project?
+    if ( QgsMapLayer *layer = u.mapLayer() )
+    {
+      for ( int i = 0; i < mModel->rowCount(); ++i )
+      {
+        // try to match project layers to current layers
+        QString userRole = mModel->item( i )->data( Qt::UserRole ).toString();
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:415]:
```suggestion
        const QString userRole = mModel->item( i )->data( Qt::UserRole ).toString();
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 16]
```diff
@@ -341,6 +400,64 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+  }
+  return indexes;
+}
+
+void QgsProcessingMultipleInputPanelWidget::dragEnterEvent( QDragEnterEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:431]:
```suggestion
  const QList< int> indexes = existingMapLayerFromMimeData( event->mimeData() );
```

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 17]
```diff
@@ -341,6 +400,64 @@ void QgsProcessingMultipleInputPanelWidget::addDirectory()
...
+    event->accept();
+  }
+}
+
+void QgsProcessingMultipleInputPanelWidget::dropEvent( QDropEvent *event )
+{
+  if ( !( event->possibleActions() & Qt::CopyAction ) )
+    return;
+
+  const QList< int> indexes =  existingMapLayerFromMimeData( event->mimeData() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:445]:
```suggestion
  const QList< int> indexes = existingMapLayerFromMimeData( event->mimeData() );
```

**Comment by roya0045** [comment]
@nyalldawson Thanks Nyall

**Pull request merged**